modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -82,14 +83,16 @@
  * Abstract base for scrolling across a search and executing bulk actions on all results. All package private methods are package private so
  * their tests can use them. Most methods run in the listener thread pool because the are meant to be fast and don't expect to block.
  */
-public abstract class AbstractAsyncBulkByScrollAction<Request extends AbstractBulkByScrollRequest<Request>> {
+public abstract class AbstractAsyncBulkByScrollAction<Request extends AbstractBulkByScrollRequest<Request>,
+    Action extends TransportAction<Request, ?>> {
```

Comment:
The introduction of the `Action` as a generic type parameter and field (`mainAction` below) is an attempt to solve the problems that come from the way `AbstractAsyncBulkByScrollAction` calls `buildScrollableResultSource` from its constructor.
Because the call is made from the superclass's constructor, it means that `TransportReindexAction.AsyncIndexBySearchAction.buildScrollableResultSource` (and by extension `buildRestClient`) can't make use of any fields in the subclass.
But `buildRestClient` needs access to the `ReindexSslConfig` instance, so we either need to push that up into the base class, or come up with a different approach.
By pushing a generic `mainAction` up into the base class, we can access fields from the `Action` without needing to duplicate them in `AbstractAsyncBulkByScrollAction`. This allows us to also remove `scriptService` from this class.